CMakeLists.txt

```diff
@@ -1,31 +1,20 @@
-cmake_minimum_required(VERSION 2.8.3)
-project(serial)
+cmake_minimum_required(VERSION 3.5.0)
 
-# Find catkin
-find_package(catkin REQUIRED)
+# General setup
+option(BUILD_SHARED_LIBS "Build using shared libraries" ON)
+
+set(PROJ_SOVERSION 1)
```

Comment:
Why have this variable instead of just hardcoding this value a few lines down where it's used?

Reply:
Similar to the recommendation to not use the `PROJECT_NAME` variable, biasing away from creating variables in cmake files makes them more boring and easier to read. CMake is a hard language to read and get correct and the more boring you can make it the nicer you are to yourself and others in the future.

One example where I've gone the other way from this recommendation though is in ros2_control with our use of the `THIS_PACKAGE_INCLUDE_DEPENDS` variable. You can find an example of that here: https://github.com/ros-controls/ros2_controllers/blob/f059022622e12115d006e0677ca2e312eed78858/joint_trajectory_controller/CMakeLists.txt#L8
The way I justify that variable is it greatly shortens the cmake file and cuts down on copy-paste errors in the dependencies that are specified for the project. I do not think your use of `PROJ_SOVERSION` is that case because it is just the number 1 and using the variable makes the cmake file longer (not to count the declaration of this variable) and adds indirection in every place where you use it. Is this a helpful explanation?

Reply:
Makes sense. I will remove the PROJ_SOVERSION and just use 1.

all of the PROJECT_NAME variables come from cherry-pick from https://github.com/wjwwood/serial/pull/231 and a later change in that PR makes it possible to have an alternative name for this library `cxx-serial`  https://github.com/wjwwood/serial/pull/231/commits/5f1ab387b7dd7bc9f08bd5d8961a7e101a5dcca2 which is what @leamas released into the debian packages.

There are so many diverging forks of this library :'(

But I agree and will change PROJECT_NAME back to serial for readability in a separate commit so if @wjwwood or @leamas chime in to this discussion it could always be reverted.

Reply:
PROJ_SOVERSION is actually crucial. It defines the binary compatibility level, and as such has been 1 so far. However, all serious packaging relies on a defined SOVERSION, without it all dependent packages will need to be recompiled for each and every change in the this lib, also the ones which are downwards compatible.

Reply:
That the library actually needs a new name is a reason, perhaps the only compelling one,  to make a friendly fork.  Forking is of course a bad thing, but perhaps the correct way to get out of the rabbit hole created by the  too general serial/serial.h.  It would make it easier to make big change while keeping the current lib as-is, for sure.

Issues could be moved to a fork without too much hassle, either in the GUI one-by-one or on the command line using a bulk transfer.

Reply:
@leamas Nobody is asking to stop specifying the SOVERSION. We're saying that we don't need an extra variable to hold the value `1`. That value can be hardcoded in the `set_target_properties` call where it's needed.

Reply:
Keeping the name is about traceability. Put it this way: if you replace SOVERSION with a an explicit constant like 1 or 2 it will need a comment explaining what it's all about. Such needs for comments is a sure sign of bad design

Yes, cmake is a from time to time a hard language. However, this makes is even more important to stick to standard programming principles, one of which being to declare constants rather than using literal values. That is, SOVERSION is to be preferred to literal whatever. 

Would to propose a similar change to a C/C++ program? If you did, what do think the reaction would be?

Reply:
How is
```cmake
set_target_properties(${PROJECT_NAME} PROPERTIES
     VERSION ${PROJECT_VERSION}
     SOVERSION ${PROJ_SOVERSION}
     PUBLIC_HEADER "${serial_HEADERS}"
 )
 ```
 better than
 ```cmake
 set_target_properties(${PROJECT_NAME} PROPERTIES
     VERSION ${PROJECT_VERSION}
     SOVERSION 1
     PUBLIC_HEADER "${serial_HEADERS}"
 )
 ```
 
 How is `SOVERSION 1` unclear whatsoever? No comments are required.